### ISSUE TITLE
Training/fine-tuning code

### ISSUE THREAD

**Issue by thivux** [kyutai-labs/pocket-tts#30]
hi team, thank you for the awesome work. i see in the paper that you say you'll release the training code. can you please let us know the timeline for that?

**Comment by gabrieldemarmiesse** [comment]
For the moment we don't have any timeline on the subject, notably because it's a lot of work to release training code. If someone wants to try to make some public training code before us though, feel free to ask questions here. There is no secret sauce and everything was detailed in the paper and the blog post. We'd be happy to help you out by giving information if you're stuck.

**Comment by dnhkng** [comment]
If you want to dump messy training code, I would be happy to tidy it up. This would be [perfect to use in my project](https://github.com/dnhkng/GLaDOS), but I would need to fine-tune the voices, as I currently have to use two TTS inference model, both Koroko and a custom trained Piper voice.

I would try and refactor using PyTorch Lightning; if you could share the main model and loss loops, I would give it a go. I don't require all the other data processing code (which can get pretty messy!). Just share the own or two key files, and I'll do the rest.

**Comment by vvolhejn** [comment]
As Gabriel said, we are not planning to release training or fine-tuning code, at least in the short term. It's more work than one might expect and serves a fairly small portion of the user base: only a fraction of people _using_ Pocket TTS will have the knowledge, resources, and time to train their own. We prefer to focus on work that will benefit a wider segment of our users.

We are looking into training the model on more languages ourselves, stay tuned!

**Comment by dnhkng** [comment]
@vvolhejn @gabrieldemarmiesse 

No worries. The model uses a 32-d latent space, but only output_proj (32→512) is provided. Could you release the encoder (512→32) used during training? Without it, I can't create valid training targets.

I have tried training one (by attempting to generate a dataset of latents, using a MLP), but it didn't work, and the fine-tuned generated garbage audio.

**Comment by vvolhejn** [comment]
@dnhkng 
https://drive.google.com/file/d/1WP7RiUh5Niiz0na12vCSRuM1E7YGZYfQ/view?usp=drive_link Here is the input projection weight, hope this helps!

**Comment by dnhkng** [comment]
This is overnight training, starting with Eponine:

[step18400_sample2.wav](https://github.com/user-attachments/files/25017420/step18400_sample2.wav)
[step18400_sample3.wav](https://github.com/user-attachments/files/25017422/step18400_sample3.wav)

It's quite tricky top fine tune.  I can get OK results with training everything except for the Mimi codec, but I can't reach the quality I would like. If I train the Mimi codec too, the model starts generating garbage (as you can hear in step18400_sample3.wav. Here are the samples from my trained Piper model in comparison:

[baseline_sample2.wav](https://github.com/user-attachments/files/25017505/baseline_sample2.wav)
[baseline_sample3.wav](https://github.com/user-attachments/files/25017511/baseline_sample3.wav)

**Comment by gabrieldemarmiesse** [comment]
Related: https://github.com/kyutai-labs/pocket-tts/issues/118

**Comment by vodp** [comment]
@dnhkng great work! I think Mimi has to be independently trained, not jointly.

**Comment by vodp** [comment]
@vvolhejn  @dnhkng  Can you share a bit more of how audio data is preprocessed for training/finetuning. Is there any normalization/standardisation applied to raw waveform ?

**Comment by manukyutai** [comment]
Mimi is indeed trained independently.

If you normalize, your model will be more robust but may produce more diverse volumes. I didn't apply explicit normalization mechanisms but it may be that the public datasets I was using already underwent some normalization processes.

**Comment by dnhkng** [comment]
I am fine tuning on a single voice (GLaDOS). I get amazing results fast (minutes of training), but I have a huge issue with reverb. 

It is a robot voice. I don't know if the issue is with Mimi, or the encoder.

**Comment by vodp** [comment]
One thing I've noticed when writing training code is the `bos_emb` (is it learnable ? ), and `emb_std` and `emb_mean` to normalize and de-normalize  (audio?) latents after mimi encoding and before mimi decoding. 
I don' know how `emb_std` and `emb_mean` are calculated, so I approximate it with a decay factor when new training data enter forward() pass, like this. Would that work ?

```
    @torch.no_grad()
    def _update_emb_stats(self, raw_latents: torch.Tensor, mask: torch.Tensor):
        """Update emb_mean and emb_std buffers via EMA over valid frames.

        Args:
            raw_latents: Raw latents from encode_audio [B, S, D]
            mask: Valid frame mask [B, S] (1.0 for valid, 0.0 for padding)
        """
        decay = self.config.emb_stats_ema_decay
        mask_3d = mask.unsqueeze(-1)  # [B, S, 1]
        num_valid = mask_3d.sum()

        if num_valid < 2:
            return

        batch_mean = (raw_latents * mask_3d).sum(dim=(0, 1)) / num_valid
        diff = raw_latents - batch_mean.unsqueeze(0).unsqueeze(0)
        batch_var = (diff * diff * mask_3d).sum(dim=(0, 1)) / (num_valid - 1)
        batch_std = batch_var.sqrt().clamp(min=1e-6)

        self.flow_lm.emb_mean.mul_(decay).add_(batch_mean, alpha=1 - decay)
        self.flow_lm.emb_std.mul_(decay).add_(batch_std, alpha=1 - decay)
```

**Comment by mananchawla2005** [comment]
can someone make it open source?

**Comment by Aratako** [comment]
This isn't exactly pocket-tts fine-tuning, but I gave a shot at reproducing the original CALM (Continuous Audio Language Models) training code and published it.                      

https://github.com/Aratako/CALM-DACVAE                                                                

I'm using DACVAE instead of Mimi as the audio VAE, so it's not a direct pocket-tts reproduction.
However, the rest of the implementation (architecture, LSD training, inference, etc.) follows the CALM paper and pocket-tts as closely as possible.

Training itself works fine and loss decreases steadily, but I haven't been able to get the model to produce intelligible speech with the data and compute I had available. It's unclear whether this is due to bugs in my code or simply insufficient model size / training data — DACVAE's latent dimension (128) is much larger than Mimi's (32), so it likely needs a larger model to handle the richer latent space.

A bit different from what's being discussed here, but hopefully it's useful as a reference. There's a good chance my code has bugs or mistakes, so any feedback would be very appreciated!

**Comment by vodp** [comment]
@Aratako You meant output audio is gibberish ?

**Comment by vodp** [comment]
@manukyutai @vvolhejn Can you help suggesting a bit more of the adaptive weighting function mentioned in Appendix A ? The implementation that @Aratako provided is kinda right direction, however LSD problem in that paper is quite small size compare to Pocket-tts. Even when I disabled LSD loss so only FM loss is left, the gradients seem totally absorbed by the adaptive weighting function, than focusing on minimizing the L2 loss term.

<img width="1642" height="1143" alt="Image" src="https://github.com/user-attachments/assets/1aeadd93-965c-4952-bbb4-f4bd932655fc" />

**Comment by vvolhejn** [comment]
We are not planning to release fine-tuning code for our TTS and STT models in the near future (Pocket TTS, Kyutai TTS 1.6B, and Kyutai STT). Since this is something that comes up repeatedly, let me expand on our reasoning here:

**It’s useful to a small fraction of people.** For example, Pocket TTS can be used by anyone with a laptop. But training one of our models on a new language requires a lot of data, compute, and ML experience. If you are in the fortunate position of having all of these, creating your own training script should not be a big hurdle – we describe our research in detail in our academic papers ([CALM](https://arxiv.org/abs/2509.06926), [DSM](https://arxiv.org/abs/2509.08753), [Moshi](https://arxiv.org/abs/2410.00037)).

**Kyutai primarily does research, not products.** This is a tricky one because the boundary between ML research and products is always fuzzy, but our primary goal is to make scientific contributions such as [delayed streams modeling](https://arxiv.org/abs/2509.08753) and [continuous audio language models](https://arxiv.org/abs/2509.06926). Of course, we aim to release models that are useful to the community, but we are a small team and we need to prioritize carefully. We maintain the inference repos for [Pocket TTS](https://github.com/kyutai-labs/pocket-tts), [DSM](https://github.com/kyutai-labs/delayed-streams-modeling/) and [Unmute](https://github.com/kyutai-labs/unmute/) and monitor the issues to help out our users. But sometimes we need to choose between working on the next research project and making our existing models more production-ready: things like fine-tuning scripts, more languages or emotion tags are usually too far towards the production side to be prioritized. For production-grade models that build on our open research, you can check out [Gradium](https://gradium.ai/).

**Comment by thewh1teagle** [comment]
@vvolhejn Thanks for the detailed explanation! I completely understand the prioritization and I appreciate all the open research you've already released.

From my perspective though, it can take me as little as ~5 hours to fine tune a new language once the pieces are in place. If there aren't other reasons for keeping them private, I'd really appreciate having the pretrained weights together with a minimal fine tuning script, so I could simply continue training on my own dataset to add support for another language.

I know Kyutai plans to add more languages over time, but in the meantime many people can add support for their own language if they have the data and the necessary hardware.

Also, a basic "handoff" fine tuning script is honestly a very small amount of work nowadays. Claude Code could probably generate a working single file training script in a few minutes. So I don't really see a strong reason not to release the weights on Hugging Face together with a simple FT script... unless there are other reasons that aren't mentioned here, of course.

**Comment by freds0** [comment]
I "vibe-coded" the fine-tuning code for pocket-tts; I tested it with LJSpeech and it worked correctly, as well as in Portuguese using a dataset in the same format as LJSpeech. For anyone who wants to test it, here is the code:


[https://github.com/freds0/pocket-tts](https://github.com/freds0/pocket-tts)

**Comment by manukyutai** [comment]
The training code is out! Have fun making your own pocket tts

**Comment by LeoMerlino** [comment]
Is model fine-tuning available?

**Comment by manukyutai** [comment]
Yes, we published a 24l model English to finetune from if necessary

**Comment by thewh1teagle** [comment]
Thanks! FWIW I got IPA input working with a LoRA fine-tune: I kept the `SentencePiece` table frozen and appended ~130 new char-level tokens (`<IPA_U0283>` style, one per IPA char), warm-started from similar grapheme pieces, then trained just those rows first and added a small LoRA (`r=8-16` on attention) + out_eos after. With only ~10h of Hebrew it speaks Hebrew with decent prosody surprisingly fast.

The one thing I couldn't solve: preserving voice cloning. Once the backbone trains at all (even LoRA-only), the new language gets entangled with the training voices. A Hebrew sentence with an unseen reference drifts toward the training timbre. Q/K-only LoRA and L2-anchoring hidden states to the frozen base reduce it but not fully, and they cost prosody.

Do you have an idea which modules/recipe would let pronunciation+prosody adapt while keeping the voice-clone pathway intact (freeze speaker_proj_weight? specific layers? voice-prompt sampling / CFG dropout tricks)? Would make low-resource language fine-tunes really easy.
